frame/support/src/storage/mod.rs

```diff
@@ -43,7 +85,16 @@ pub fn with_transaction<R>(f: impl FnOnce() -> TransactionOutcome<R>) -> R {
 	use TransactionOutcome::*;
 
 	start_transaction();
-	match f() {
+
+	#[cfg(all(feature = "std", debug_assertions))]
+	debug_helper::inc_transaction_level();
+
+	let res = f();
+
+	#[cfg(all(feature = "std", debug_assertions))]
+	debug_helper::dec_transaction_level();
```

Comment:
If `f()` is panicking, this will never be called. While we can argue if that is really required for tests, I think that we should make it right anyway.

Make `inc_transaction_level` return some guard, that when dropped, will decrease the transaction level.